### PR TITLE
Add sample support of `YOLOv9` and `YOLOv10` in OpenCV

### PR DESCRIPTION
**Pull request by Abdurrahheem**
This PR adds sample support of  [`YOLOv9`](https://github.com/WongKinYiu/yolov9) and [`YOLOv10`](https://github.com/THU-MIG/yolov10/tree/main)) in OpenCV. Models for this test are located in this [PR](https://github.com/opencv/opencv_extra/pull/1186). 

**Running YOLOv10 using OpenCV.** 
1. In oder to run `YOLOv10` one needs to cut off postporcessing with dynamic shapes from torch and then convert it to ONNX. If someone is looking for ready solution, there is [this forked branch](https://github.com/Abdurrahheem/yolov10/tree/ash/opencv-export) from official YOLOv10.  Particularty follow this proceduce. 

```bash
git clone git@github.com:Abdurrahheem/yolov10.git
conda create -n yolov10 python=3.9
conda activate yolov10
pip install -r requirements.txt
python export_opencv.py --model=<model-name> --imgsz=<input-img-size>
```
By default `model="yolov10s"` and `imgsz=(480,640)`. This will generate file `yolov10s.onnx`, which can be use for inference in OpenCV

2. For inference part on OpenCV.  one can use `yolo_detector.cpp` [sample](https://github.com/opencv/opencv/blob/4.x/samples/dnn/yolo_detector.cpp). If you have followed above exporting procedure, then you can use following command to run the model. 

``` bash
build opencv from source 
cd build 
./bin/example_dnn_yolo_detector --model=<path-to-yolov10s.onnx-file> --yolo=yolov10 --width=640 --height=480 --input=<path-to-image> --scale=0.003921568627 --padvalue=114
```
If you do not specify `--input` argument, OpenCV will grab first camera that is avaliable on your platform. 
For more deatils on how to run the `yolo_detector.cpp` file see this [guide](https://docs.opencv.org/4.x/da/d9d/tutorial_dnn_yolo.html#autotoc_md443) 

**Running YOLOv9 using OpenCV**

1. Export model following [official guide](https://github.com/WongKinYiu/yolov9)of the YOLOv9 repository. Particularly you can do following for converting.

```bash
git clone https://github.com/WongKinYiu/yolov9.git
cd yolov9
conda create -n yolov9 python=3.9
conda activate yolov9
pip install -r requirements.txt
wget https://github.com/WongKinYiu/yolov9/releases/download/v0.1/yolov9-t-converted.pt
python export.py --weights=./yolov9-t-converted.pt --include=onnx --img-size=(480,640) 
```

This will generate <yolov9-t-converted.onnx> file.

2.  Inference on OpenCV.

```bash
build opencv from source 
cd build 
./bin/example_dnn_yolo_detector --model=<path-to-yolov9-t-converted.onnx> --yolo=yolov9 --width=640 --height=480 --scale=0.003921568627 --padvalue=114 --path=<path-to-image>
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/dnn/yolo_detector.cpp [hunk 1]
```diff
@@ -131,25 +133,28 @@ void yoloPostProcessing(
         // remove the second element
         outs.pop_back();
         // unsqueeze the first dimension
-        outs[0] = outs[0].reshape(0, std::vector<int>{1, 8400, 84});
+        outs[0] = outs[0].reshape(0, std::vector<int>{1, 8400, nc + 4});
     }
 
+    // assert if last dim is 85 or 84
+    CV_CheckEQ((outs[0].size[2] == nc + 5 || outs[0].size[2] == 80 + 4), true, "Invalid output shape: ");
```
Comment (dkurt) [inline samples/dnn/yolo_detector.cpp:140]: Check also that `outs[0].dims == 2` before.

File: samples/dnn/yolo_detector.cpp [hunk 2]
```diff
@@ -131,25 +133,28 @@ void yoloPostProcessing(
         // remove the second element
         outs.pop_back();
         // unsqueeze the first dimension
-        outs[0] = outs[0].reshape(0, std::vector<int>{1, 8400, 84});
+        outs[0] = outs[0].reshape(0, std::vector<int>{1, 8400, nc + 4});
```
Comment (dkurt) [inline samples/dnn/yolo_detector.cpp:136]: 8400 is also depends on number of classes? So should it be `(nc + 4) * 100` or not?

Comment (dkurt) [reply samples/dnn/yolo_detector.cpp:136]: `nc` seems redundant as it can be computed from original shape:
```cpp
int nc = outs[0].total() / 8400 - 4;
```
or, if 8400 is also depends on `nc`:
```cpp
int nc = sqrt(outs[0].total() / 100) - 4;
```

Comment (Abdurrahheem) [reply samples/dnn/yolo_detector.cpp:137]: > 8400 is also depends on number of classes? So should it be `(nc + 4) * 100` or not?

8400 depends on number of anchors only and has nothing to do with `nc`. 
`nc` depends on which dataset detector what trained. On `COCO` dataset `nc = 80`. For other datasets it might be different.

Comment (dkurt) [reply samples/dnn/yolo_detector.cpp:137]: Got it, thanks! So do you mind to compute `nc` from total number of `outs[0]` elements?

Comment (Abdurrahheem) [reply samples/dnn/yolo_detector.cpp:138]: I strongly believe that it should stay as papermeter. As this is a parameter during traning. Computing it via `outs[0]` is kind of hack.

**Comment by asmorkalov** [comment]
opencv_extra PR is missing.

File: samples/dnn/yolo_detector.cpp (hunk 2 above)
Comment (dkurt) [reply samples/dnn/yolo_detector.cpp:138]: The corner case is a background class. Some training utils may include this as a separate class - some not. So even if you trained model for 80 real classes, there is a variation nc=80 nc=81 and it may confuse even more.

Comment (Abdurrahheem) [reply samples/dnn/yolo_detector.cpp:138]: Computing `nc` using number of anchor points is not good since number of anchors depens on the image size. Inference can happen on different image resolutons, depending on use case.

**Comment by Abdurrahheem** [comment]
> opencv_extra PR is missing.

Just added. [Here](https://github.com/opencv/opencv_extra/pull/1186)

**Comment by asmorkalov** [comment]
Test and sample changes look good for me. What about Yolov10 export steps? I expect significant changes in the tutorial text.

**Comment by Abdurrahheem** [comment]
> Test and sample changes look good for me. What about Yolov10 export steps? I expect significant changes in the tutorial text.

Exprot steps are outline in[ this fork of yolov10](https://github.com/Abdurrahheem/yolov10/blob/ash/opencv-export/export_opencv.py) in GitHub. And the instruction for this are lacted in this [READme](https://dl.opencv.org/models/yolov10/)

**Comment by asmorkalov** [comment]
I propose to extend "PyTorch Model Export" for Yolov10 as soon as it's not trivial.

**Review by asmorkalov — APPROVED** [review]

**Comment by rmsalinas** [comment]
Hi. results differ significantly (worse) from the original model. 

Original model :

yolo predict model= models/yolov10s.pt  source= ultralytics/assets/bus.jpg

_Ultralytics YOLOv8.1.34 🚀 Python-3.10.14 torch-2.0.1+cu117 CPU (13th Gen Intel Core(TM) i7-13700H)
YOLOv10s summary (fused): 293 layers, 8096880 parameters, 86880 gradients, 24.8 GFLOPs

image 1/1 /home/salinas/yolov10/ultralytics/assets/bus.jpg: 640x480 4 0s, 1 5, 1 11, 81.1ms
Speed: 1.4ms preprocess, 81.1ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 480)
Results saved to /home/salinas/yolov10/runs/detect/predict5_

bus

Proposed model:
./example_dnn_yolo_detector --model=/home/salinas/yolov10_opencv/yolov10-ash-opencv-export/yolov10s.onnx -i=/home/salinas/yolov10/ultralytics/assets/bus.jpg --yolo=yolov10 --width=640 --height=480 --scale=0.003921568627 --padvalue=114  --thr=0.1

Yolo Object Detector_screenshot_09 08 2024
